### ISSUE TITLE
Enabling all signature algorithms by default.

### ISSUE THREAD

**Issue by xvzcf** [open-quantum-safe/openssl#282]
In [ssl/ssl_local.h](https://github.com/open-quantum-safe/openssl/blob/OQS-OpenSSL_1_1_1-stable/ssl/ssl_local.h#L207), there are `#define`s that assign one-hot encodings to each signature algorithm. These encodings are used in [ssl/ssl_cert_table.h](https://github.com/open-quantum-safe/openssl/blob/OQS-OpenSSL_1_1_1-stable/ssl/ssl_cert_table.h#L26), and are assigned to a uint32_t variable called `amask` [here](https://github.com/open-quantum-safe/openssl/blob/OQS-OpenSSL_1_1_1-stable/ssl/ssl_local.h#L1924).

Since the encodings are one-hot, at most 32 different signature algorithms can be encoded, and since 8 bits are already used up by OpenSSL by default, at most 24 different OQS signature algorithms can be encoded and assigned to `amask` at any given instant. 

This (possibly?) prevents us from enabling all signature algorithms by default, and the question is how we should deal with this.

**Comment by baentsch** [comment]
TLDR: After checking the code I don't see a problem going beyond the possible `amask` bitcount (32) with OQS algorithms, i.e., activate all OQS-sigs. Minor caveat on clients not sending sigalgs: See below: Ideas how to trigger that welcome.

Full rationale:

Source code analysis (code review plus walk-through by debugger) of all places where `amask` is utilized (discarding/assuming as irrelevant its uses in `test` and its own declaration in `ssl_local.h`):

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/ssl_ciph.c#L2161
-> Sole use in auth disablement check: not applicable to OQS-OpenSSL as `disabled_auth_mask` is only ever set to legacy auth algorithms

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/statem/statem_clnt.c#L1959-L1960

--> use guarded for !TLS1_3: not applicable to OQS-OpenSSL (running only TLS1.3)

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/statem/statem_clnt.c#L3606

--> ECDSA only: not applicable to OQS-OpenSSL

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/statem/statem_clnt.c#L3597-L3598

--> No cipher alg is associated with an OQS auth alg; thus, this case cannot trigger (in testing for example cipher is always `TLS_AES_256_GCM_SHA384` ensuring this code path is not reached at all) -> also no issue

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/t1_lib.c#L1325

--> _seems_ OK, but cannot validate 100% as I do not know how to enforce the client to NOT send a list of supported sig types.

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/t1_lib.c#L2231

--> Again, only operates on legacy algs in `disabled_mask`: Not applicable to OQS-OpenSSL

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/t1_lib.c#L3166

--> This code/function `tls12_get_cert_sigalg_idx` would only trigger on TLS < 1.3 as per this:

https://github.com/open-quantum-safe/openssl/blob/9e342427ead2945c7a757e8497d3d1cb7ae6d578/ssl/t1_lib.c#L3333

--> Again, not applicable to OQS-OpenSSL

**Comment by prchander** [comment]
Is this the reason why dilithium family errors out? It says the algorithm is not enabled when generate.ylm says it is true.

**Comment by baentsch** [comment]
> Is this the reason why dilithium family errors out? It says the algorithm is not enabled when generate.ylm says it is true.

I'm not sure I understand either statement: How/why does dilithium error out? What (command?) says it's not enabled?

**Comment by prchander** [comment]
Oh I am sorry I thought these my problem was related to this. "Since the encodings are one-hot, at most 32 different signature algorithms can be encoded, and since 8 bits are already used up by OpenSSL by default, at most 24 different OQS signature algorithms can be encoded and assigned to amask at any given instant." I thought that was referring to some algorithms not being enabled. I will create a new issue sorry I jumped in this thread.

**Comment by baentsch** [comment]
After having just enabled all algs, this is the (not entirely surprising) response by a self-respecting compiler (OSX):

```
In file included from ssl/ssl_cert.c:24:
ssl/ssl_cert_table.h:100:45: error: integer literal is too large to be represented in any integer type
    {EVP_PKEY_P384_SPHINCSHARAKA192FSIMPLE, SSL_aP384_SPHINCSHARAKA192FSIMPLE}, /* SSL_PKEY_P384_SPHINCSHARAKA192FSIMPLE */
                                            ^
ssl/ssl_local.h:355:43: note: expanded from macro 'SSL_aP384_SPHINCSHARAKA192FSIMPLE'
#define SSL_aP384_SPHINCSHARAKA192FSIMPLE 0x400000000000000000000U
```
--> We'd need to "widen" integers beyond 64 bits, it seems. Doesn't feel right. 

What'd be the rationale behind including all algs? Do we _really_ need _all_? The count of Sphincs' algs alone made me dizzy....

**Comment by xvzcf** [comment]
Including all algs would mean that CI runs do test every single algorithm; right now, we manually enable and test all the algorithms before a release. It would also mean that users wouldn't have to go through the algorithm enabling process.

**Comment by baentsch** [comment]
> It would also mean that users wouldn't have to go through the algorithm enabling process.

Do we know users that need _all_ these algs? On which platforms? Wouldn't it be a better approach to have a "representative" set of algorithms enabled (similar to what we have now) and do "the full monty" with all algs only on release? All it'd need for the latter is running `sed` over the checked-in "representative" `generate.yml` to save us any manual effort _and_ save the CO2 emitted by CCI doing over-and-over the same tests.

As a concrete proposal: What about disabling by default all algs that don't truly "teach something new"? For example, why have SHAKE _and_ AES _and_ plain ref algs for the same family and NIST level? What about asking all algorithm "owners" about a priority list and then using that to curtail each family at the same count (say 1 alg per NIST level)?

**Comment by xvzcf** [comment]
> Do we know users that need _all_ these algs? On which platforms? 

Not that I'm aware of, I just see this as a secondary nice-to-have. 

> Wouldn't it be a better approach to have a "representative" set of algorithms enabled (similar to what we have now) and do "the full monty" with all algs only on release? 

There's still a small chance we might not catch certain errors until release testing time, but this should be low as long as the algorithms are representative enough (which I've tried to ensure here and in OpenSSH). 

> As a concrete proposal: What about disabling by default all algs that don't truly "teach something new"? For example, why have SHAKE _and_ AES _and_ plain ref algs for the same family and NIST level? What about asking all algorithm "owners" about a priority list and then using that to curtail each family at the same count (say 1 alg per NIST level)?

When I took a shot at something like this last year, it was not well received, but we can revisit this.

**Comment by baentsch** [comment]
> Not that I'm aware of, I just see this as a secondary nice-to-have.

Then I'd say we take our responsibility for the environment more serious and run _all_ algs as rarely as reasonable.

> When I took a shot at something like this last year, it was not well received, but we can revisit this.

So maybe a topic for our call (@dstebila ?)

**Comment by baentsch** [comment]
Agreement to leave as-is for now & extend release testing along the lines discussed above.
